### DISCUSSION TITLE
How to redifine Query & Mutation type for multiple schema ?

### DISCUSSION THREAD

**Discussion by callmephil** [heftapp/graphql_codegen#272] [Q&A] [answered]
Hey ! 

I have multiple schemas (files) and each one of them has Query & Mutation Type declared in them.

Example: 

Magic.graphql: 
```
type Magic {
  id: ID!
}

input MagicCreateInput {
  id: ID!
}

type Query {
  magic(): Magic
}

type Mutation {
  createMagic(input: MagicCreateInput!): Magic!
}

```

Power.graphql
```
type Power {
  id: ID!
}

input PowerCreateInput {
  id: ID!
}

type Query {
  power(): Power
}

type Mutation {
  createPower(input: PowerCreateInput!): Power!
}
```

When I'm running build_runner the code gen will not generate Query & Mutation for Power
and the following Invalid error is thrown 
> [SEVERE] GraphQL: Failed to find type for field Power on Query

I'm not sure if that is supported yet. But if it is anyone knows what I missed?

**Comment by budde377** [comment]
Take a look at scopes: https://github.com/heftapp/graphql_codegen/tree/main/packages/graphql_codegen#multiple-schemas

**Reply by callmephil** [reply]
Thanks for replying! 

The scope would work if the schemas are totally independent of each other, Here's a more complex case where type & Fragments would depend on each other:

./School/schema_school.graphql
```
Type School {
   id: ID!
   name: String!
}

input CreateSchoolInput {
  id: ID!
  name: String!
}

type Query {
  schools(): [School!]!
}

type Mutation {
  createSchool(input: CreateSchoolInput!}): School!
}
```
./Magic/queries_power.graphql
```
Fragment fullSchool on Power {
  id
}
```


./Magic/schema_magic.graphql
```
type Magic {
  id: ID!
  school: School!
}

input MagicCreateInput {
  id: ID!
}

type Query {
  magic(): Magic
}

type Mutation {
  createMagic(input: MagicCreateInput!): Magic!
}
```
./Magic/schema_power.graphql
```
type Power {
  id: ID!
  schools: [School!]!
}

input PowerCreateInput {
  id: ID!
}

type Query {
  power(): Power
}

type Mutation {
  createPower(input: PowerCreateInput!): Power!
}
```
./Magic/queries_power.graphql
```
Fragment fullMagic on Power {
  id
  school { ...fullSchool }
}
```

Now both Magic & Power are dependent on School, but School is scoped so it isn't being recognized by the generator,

This is the output of build_runner:
```
[SEVERE] graphql_codegen on lib/gql/power/schema_power.graphql:
Invalid GraphQL: Failed to find type-definition for field name
```

**Reply by budde377** [reply, marked as answer]
This seems like a very strange use-case. In which cases would this happen? You can take a look at `extension` which will allow you to extend the `Query` in your example.

**Reply by callmephil** [reply]
Thanks! extensions worked =)

I have found the issue. I was trying to extend Query & Mutation type on my schemas but I didn't had defined the type to start with. So I defined them in the schema.graphql which I use as common for scalar and custom types.

So..

Schema.graphql declare type Query {} and type Mutation {}

Power, Magic and school extend Query { .. } and extend Mutation { ... }

And my folder layer is as follow :
Queries (files..)
Fragments  (files..)
Mutations  (files..) 
Schemas  (files..)

It would be nice to add to the documentation this might be helpful for people that build a very large application that requires to re-use the code as much as possible =)

**Reply by budde377** [reply]
A PR with documentation changes is always highly appreciated
